tests/test_denest.py

```diff
@@ -184,3 +184,87 @@ def test__schema__nested_arrays_add_tables():
 
         print('PASSED')
         print()
+
+
+def test__records__nested():
+    denested = denest.to_table_batches({
+        "properties": {
+            "a": {"type": "object",
+                  "properties": {
+                      "b": {
+                          "type": "array",
+                          "items": {
+                              "type": "object",
+                              "properties": {
+                                  "c": {
+                                      "type": "object",
+                                      "properties": {
+                                          "d": {"type": "integer"},
+                                          "e": {"type": "array",
+                                                "items": {"type": "object",
+                                                          "properties": {
+                                                              "f": {"type": "string"},
+                                                              "g": {"type": "boolean"}}}}}}}}}}}}},
+        [],
+        [{"a": {"b": []}},
+         {"a": {"b": [{"c": {"d": 1}}]}},
+         {"a": {"b": [{"c": {"d": 12}},
+                      {"c": {"d": 123}}]}},
+         {"a": {"b": [{"c": {"d": 1234}},
+                      {"c": {"d": 12345}},
+                      {"c": {"d": 123456}}]}},
+         {"a": {"b": [{"c": {"e": [{"f": "hello",
+                                    "g": True},
+                                   {"f": "goodbye",
+                                    "g": True}]}}]}}])
+
+    print('denested:', denested)
+
+    assert 3 == len(denested)
+
+    root_table_checked = False
+    nested_table_checked = False
+    nested_nested_table_checked = False
+
+    for table_batch in denested:
+        if tuple() == table_batch['streamed_schema']['path']:
+            root_table_checked = True
+
+            assert {} == table_batch['streamed_schema']['schema']['properties']
+
+            assert 5 == len(table_batch['records'])
```

Comment:
This is not really related to this PR:
When I was running the tests and looking at the printed denested data I started thinking about this line. Why is there empty dicts ({}) added to the records field in the `_denest_record` function? Even if the record doesn't contain any items we still add an empty dict. Is that desired?

We could change this with a simple check in `_denest_record` when appending denested records:
```python
if denested_record != {}:
        records_map[table_path].append(denested_record)
```
If we want to change this we should do it in another PR. Also, this change makes some length asserts fail and that needs to be addressed as well in that case.

Reply:
Yeah, it's necessary. So if you get rid of that, then there's no way for our code to create a key for the denested values to relate _back_ to.

ie, think of if you had a schema like this:

```json
{'properties':
  {'a': {'items': [...]},
   'b': {'items': [...]}}}
```

So `a` and `b` _relate_ to each other in that they're on the same parent right? But when we denest, this'll turn into:

```
root: {}
root__a: {...}
root__b: {...}
```

Since we generate uuids for tables which don't have `key_properties` set, we can then relate the child tables and records back to each other in postgres.

If we get rid of the record, it indicates to everything else that there's just nothing there...

(now, mind you, the code _above_ here where we're setting up the `Singer Stream` etc., actually injects a `key_properties` and uuids etc. so you can't ever have this case of empty tables. There'll always be the StitchData `_sdc` columns/fields/properties)

Reply:
Ah I see, thanks for the explanation! Makes perfect sense 👍